[FormControl] Create FormControlUnstyled

michaldudak opened the pull request:
This PR adds a new unstyled component: FormControlUnstyled. Most of the code is taken from FormControl. Props related to color, margin and other visual attributes were removed. FormControl adds these fields to the context with the `extraContextProperties` prop.

oliviertassinari commented on packages/material-ui/src/FormControl/FormControl.js:
```diff
@@ -111,119 +111,34 @@ const FormControl = React.forwardRef(function FormControl(inProps, ref) {
-        {children}
-      </FormControlRoot>
-    </FormControlContext.Provider>
+    <FormControlUnstyled
+      className={clsx(classes.root, className)}
+      component={FormControlRoot}
+      componentsProps={componentsProps}
+      disabled={disabled}
+      error={error}
+      extraContextProperties={childContext}
```
Seeing this, it makes me wonder if MD shouldn't be built on the hook API instead of the unstyled one. I'm asking because using the hook directly removes an indirection:

1. it might improve performance as we skip a React layer (we save the ...spread, React element cost overhead). 
2. it makes it easier for developers to follow our sources 

Anyway, I'm not saying we should do it, but it could be a valid option.

michaldudak replied:
This was my first idea when developing the Switch. It's actually easier to implement the MD Switch using the hooks. Plus, we lose one layer of abstraction.
The reason I used `SwitchUnstyled` instead of `useSwitch` in https://github.com/mui-org/material-ui/pull/27182 is to have higher test coverage. Since `SwitchUnstyled` uses the hook, by testing `SwitchUnstyled` we also test the hook. I also have the hook implementation ready, so once we're good to merge, we can decide which one to take. 

With FormControl it's a bit different. The existing hook doesn't encapsulate the logic and it's just a context accessor, so here it makes more sense to implement `FormControl` using `FormControlUnstyled`.

eps1lon commented on packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.tsx:
```diff
@@ -0,0 +1,257 @@
+   * @default false
+   */
+  error: PropTypes.bool,
+  /**
+   * Extra properties to be placed on the FormControlContext.
+   * @default {}
+   */
+  extraContextProperties: PropTypes.object,
```
Can be implemented via separate context. Merging context is not ok at the moment. Separate contexts don't have the same problems as merged context values.

Let's not add features before we've defined what they're used for.

michaldudak replied:
OK, do you have something like this in mind?

```jsx
<FormControl>
  <FormControlContext.Provider value={thingsSpecificToMd}>
    <FormControlUnstyled>
      <FormControlUnstyledContext.Provider value={genericThings}>
        <!-- actual form controls -->
```

and then we'd have MD-version of `useFormControl` that would call both `useContext(FormControlUnstyledContext)` and `useContext(FormControlContext)` and return their results?

mnajdova replied:
+1 for not adding this via props. The main reason is that it would be confusing if it exists on some components, but not on others. The code that you've shared @michaldudak makes sense to me.

eps1lon replied:
> OK, do you have something like this in mind?
> 
> ```js
> <FormControl>
>   <FormControlContext.Provider value={thingsSpecificToMd}>
>     <FormControlUnstyled>
>       <FormControlUnstyledContext.Provider value={genericThings}>
>         <!-- actual form controls -->
> ```

Yep 👍🏻 Everything else would be confusing since you'd have to always be aware in which kind of `FormControlUnstyledContext` you are i.e. which `value` is provided.

eps1lon commented on packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.tsx:
```diff
@@ -0,0 +1,257 @@
+  focused: PropTypes.bool,
+  /**
+   * If `true`, the label is hidden.
+   * This is used to increase density for a `FilledInput`.
```
This is a Material-UI component. It seems like this prop is unrelated to an "unstyled" version of the FormControl.

eps1lon commented on packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.tsx:
```diff
@@ -0,0 +1,257 @@
+    required,
+  };
+
+  const [adornedStart, setAdornedStart] = React.useState(() => {
```
This is also related to Material design. Looks unrelated to an "unstyled" version.

eps1lon replied:
Though I remember that we also needed to clarify what an "adornment" is and what a "suffix" is in the TextField. So there is potentially a place for a generic "adornment" (which is probably better named as "secondary action") and "suffix" (e.g. a currency symbol) in the unstyled package.

eps1lon reviewed (changes requested):
This looks ok so far but doesn't look like an "unstyled" version of the `FormControl`. It just has no styles which is not what we envision the "unstyled" package is for: It should contain primitives to build a new design system. But right now the `FormControlUnstyled` still contains concepts that are related to Material design.

Since we create a new component we have the opportunity to avoid old mistakes:
`filled` should not be a prop but derived from the `value`. The `FormControl` should receive the value instead and propagate it (and `filled`) accordingly. This avoids cascading updates from the child back to the parent i.e. we save a full commit and make sure the components relying on `FormControl` are always in a consistent state.

michaldudak commented:
> filled should not be a prop but derived from the value. The FormControl should receive the value instead and propagate it (and filled) accordingly.

👍🏻 I like the idea.
So do you think the case where a value is received directly by an input inside a FormControl should not be supported and result in an error?

eps1lon commented:
> So do you think the case where a value is received directly by an input inside a FormControl should not be supported and result in an error?

It comes down to the old "context vs props" discussion. For historic reasons we support overriding a context value via props. I don't think this is ever the right thing to do but rather misuses semantics of certain props. 

So in the end `value` should only be specified in the `FormControl` and everything else should read the context value. Just like `focused` or `disabled`.

eps1lon commented on packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.test.tsx:
```diff
@@ -21,4 +45,165 @@ describe('<FormControlUnstyled />', () => {
+      );
+
+      expect(readContext.args[0][0]).to.have.property('value', undefined);
+      expect(readContext.args[0][0]).to.have.property('defaultValue', undefined);
```
I think in this test you can use `expect(getByRole('textbox')).to.have.property('value', "")`. Can you go over the tests again and try not to use `readContext` and instead assert on the generated DOM? 

Right now you're creating a test component for two testing APIs but only make use of one. I'm trying to avoid test helpers local to a test as much as possible since hurts readability of a test. 

The implementation with `useEffect` is also slightly broken in React 18 but we have a workaround for that. But if we can stick with existing helpers, let's do it.

eps1lon commented on packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.tsx:
```diff
@@ -0,0 +1,234 @@
+  } = props;
+
+  const isControlled = value !== undefined;
+  const [filled, setFilled] = React.useState(() => isFilled(value, defaultValue));
```
`filled` shouldn't need to be a separate state since it depends on `value`. The value of `filled` can be directly derived from `value`.

michaldudak replied:
With the current implementation, it must be a separate state, as the inner input can be either controlled or uncontrolled.
But now it got to me this could be simplified in a way that even if FormControlUnstyled gets `defaultValue`, it won't propagate it to the context, but rather always set a `value`. This way `filled` could indeed be simply derived from value. Do you think this would be a better approach?

eps1lon replied:
> This way `filled` could indeed be simply derived from value. Do you think this would be a better approach?

Yep, that's the goal. The current implementation is more complicated than it actually needs to. By ensuring a consistent state we also avoid cascading updates.

eps1lon commented on packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.tsx:
```diff
@@ -0,0 +1,234 @@
+        {...rootProps}
+        className={clsx(
+          classes.root,
+          className,
+          rootProps?.className,
```
nit: do we need to handle `className` and `rootProps.className` separately? I think `rootProps.className` is sufficient if we stop destructuring `className`.

eps1lon reviewed (changes requested):
We can't write polymorphic components (i.e. the ones implementing `component`) in TypeScript. We need the separation between runtime and types: https://github.com/mui-org/material-ui/issues/15984#issuecomment-684506803

Otherwise the declaration is incomplete.

michaldudak commented:
We can't write polymorphic components (i.e. the ones implementing component) in TypeScript. We need the separation between runtime and types: #15984 (comment)

@eps1lon, out of curiosity, what makes the type definitions in d.ts files work better than typed implementation in .ts file?

eps1lon commented:
Compare the generated declaration files with our current ones and then apply them to https://github.com/mui-org/material-ui/issues/15984#issuecomment-684506803

michaldudak commented:
@eps1lon Could you take a look at https://github.com/mui-org/material-ui/pull/27240/files#diff-ac6c93d256628e694cee09f40532debb9f0482ff8f2d591f2fd458da6fe8a6a8 once again? I added a few types and a type assertion to the exported FormControlUnstyled. The generated .d.ts is now equivalent to the one I would write by hand. I also added a few type tests (based on ones I found in Paper.spec.ts).
Please let me know if I'm still missing something.

eps1lon commented on docs/translations/api-docs/form-control-unstyled/form-control-unstyled-de.json:
```diff
@@ -0,0 +1,13 @@
+    "children": "The content of the component.",
+    "disabled": "If <code>true</code>, the label, input and helper text should be displayed in a disabled state.",
+    "error": "If <code>true</code>, the label is displayed in an error state.",
+    "extraContextProperties": "Extra properties to be placed on the FormControlContext.",
```
I think you need to remove these and run `docs:api` again to get the latest state.

eps1lon commented on packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.test.tsx:
```diff
@@ -0,0 +1,200 @@
+    disabled: context?.disabled,
+    onChange: context?.onChange,
+    required: context?.required,
+    value: context?.value as any,
```
Why the type cast here?

michaldudak replied:
`context.value` has `unknown` type. I could set its type to `string | number | readonly string[]` to match the native input's props, but this would limit the usage of FormControl with other types of inputs (such as DatePicker).
Another option would be to just use `any` instead of `unknown`.

eps1lon replied:
Then let's cast it to `string` so that we understand what we're expecting here.

eps1lon commented on packages/material-ui-unstyled/src/utils/appendStyleProps.test.ts:
```diff
@@ -0,0 +1,56 @@
+  checked: true,
+};
+
+const CustomComponent = () => null;
```
Let's inline these into their specific tests. They're not used equaly and make debugging a test needlessly hard. A test suite doesn't need to follow DRY principles. A single test probably should.

eps1lon reviewed (approved):
Let's discuss the nits but overall this is good to go.

michaldudak commented:
@eps1lon to clarify - is the approach (a component in .ts, with type assertion on default export) I took here OK for future unstyled components, or should I treat it as an exception and continue in .js with other components?

eps1lon commented:
Sounds good to me. Just dont feel too safe with using TypeScript with components.
